larq_compute_engine/core/bgemm_kernels_arm.h

```diff
@@ -97,80 +66,47 @@ struct BgemmKernel<ruy::Path::kNeon, std::int8_t,
+// Optimised Aarch64 kernel with 16-bit accumulators. Supports float or int8 or
+// bitpacked output.
+template <typename DstScalar>
+struct BgemmKernel<ruy::Path::kNeon, DstScalar,
+                   BinaryMulParams<std::int16_t, DstScalar>> {
   Tuning tuning = Tuning::kAuto;
   using LhsLayout = FixedKernelLayout<Order::kColMajor, 4, 8>;
   using RhsLayout = FixedKernelLayout<Order::kColMajor, 4, 4>;
   explicit BgemmKernel(Tuning tuning_) : tuning(tuning_) {}
   void Run(const ruy::PMat<TBitpacked>& lhs, const ruy::PMat<TBitpacked>& rhs,
-           const BinaryMulParams<std::int16_t, std::int8_t>& mul_params,
+           const BinaryMulParams<std::int16_t, DstScalar>& mul_params,
            int start_row, int start_col, int end_row, int end_col,
-           ruy::Mat<std::int8_t>* dst) const {
-    BinaryKernelParams<std::int8_t, LhsLayout::kCols, RhsLayout::kCols> params;
-    MakeBinaryKernelParams(lhs, rhs, mul_params, start_row, start_col, end_row,
-                           end_col, dst, &params);
+           ruy::Mat<DstScalar>* dst) const {
+    TFLITE_DCHECK((std::is_same<DstScalar, float>::value ||
+                   std::is_same<DstScalar, std::int8_t>::value ||
+                   std::is_same<DstScalar, TBitpacked>::value));
+    BinaryKernelParams<DstScalar, LhsLayout::kCols, RhsLayout::kCols> params;
+    MakeBinaryKernelParams(lhs, rhs, start_row, start_col, end_row, end_col,
+                           dst, mul_params, &params);
     BinaryKernelNeonOutOfOrder8x4(params);
   }
 };
 
-// Fallback Aarch64 kernel for float output with 32-bit accumulators (when
-// there's a risk of overflowing 16-bit accumulators).
-template <>
-struct BgemmKernel<ruy::Path::kNeon, float,
-                   BinaryMulParams<std::int32_t, float>> {
-  Tuning tuning = Tuning::kAuto;
-  using LhsLayout = FixedKernelLayout<Order::kColMajor, 4, 4>;
-  using RhsLayout = FixedKernelLayout<Order::kColMajor, 4, 4>;
-  explicit BgemmKernel(Tuning tuning_) : tuning(tuning_) {}
-  void Run(const ruy::PMat<TBitpacked>& lhs, const ruy::PMat<TBitpacked>& rhs,
-           const BinaryMulParams<std::int32_t, float>& mul_params,
-           int start_row, int start_col, int end_row, int end_col,
-           ruy::Mat<float>* dst) const {
-    BinaryKernelParams<float, LhsLayout::kCols, RhsLayout::kCols> params;
-    MakeBinaryKernelParams(lhs, rhs, mul_params, start_row, start_col, end_row,
-                           end_col, dst, &params);
-    BinaryKernelNeonOutOfOrder4x4(params);
-  }
-};
-
-// Fallback Aarch64 kernel for int8 output with 32-bit accumulators (when
-// there's a risk of overflowing 16-bit accumulators).
-template <>
-struct BgemmKernel<ruy::Path::kNeon, std::int8_t,
-                   BinaryMulParams<std::int32_t, std::int8_t>> {
+// Fallback Aarch64 kernel with 32-bit accumulators (when there's a risk of
+// overflowing 16-bit accumulators). Supports float or int8 output.
+template <typename DstScalar>
```

Comment:
Ah nice one, I did investigate using `enable_if` in a somewhat similar way at an earlier point when developing this PR but wasn't able to get it to work properly.

Unfortunately it doesn't quite work because at some point it tries to compile this template with `TBitpacked`, then type substitution fails, and so the empty base struct gets used instead: https://github.com/larq/compute-engine/blob/18af158b4447075f6968c4acd0d30f3a13aca2df/larq_compute_engine/core/bgemm_kernels_ruy.h#L18-L19

Then at this line a proper compile error is caused: https://github.com/larq/compute-engine/blob/18af158b4447075f6968c4acd0d30f3a13aca2df/larq_compute_engine/core/bgemm_trmul_params.h#L39

because the empty base struct doesn't define a type `LhsLayout`.

I think for now let's leave it, but I might see if there's a smarter thing that we can do to solve this in #497.